### PR TITLE
NN based mobile gaze

### PR DESCRIPTION
**Pull request by tusharsangam**

### CODE REVIEW

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 1]
```diff
@@ -42,6 +42,7 @@
 from std_msgs.msg import Float32, String
 
 MAX_CMD_DURATION = 5
+TRAVLE_TIME_JOINT_BASE_ARM_CONTROL = 5  # 5 for the base; and 1/2*0.9 = 0.45 for the arm
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/base_env.py:45]: remove the comments here

File: README.md [hunk 2]
```diff
@@ -196,6 +196,7 @@ python aria_streamer.py --data-path="/path/to/directory/containing/vrs/and/mps/f
 3. `--dry-run` (Input: bool): This is used to indicate if it should command robot to move or not. If dry run is true, then it will not send the commands to the robot to make it move. *But this will still require you to have robot watching the QR code for localizing itself in the system*.
 4. `--verbose` (Input: bool): This flag is for updating the visual stream with detections from both QR & object detectors.
 
+
```
Comment (KavitShah1998) [inline README.md:199]: @tusharsangam this line change seems unncessary. Its causing an extra file change. Let's revert this.

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 3]
```diff
@@ -42,6 +42,7 @@
 from std_msgs.msg import Float32, String
 
 MAX_CMD_DURATION = 5
+TRAVLE_TIME_JOINT_BASE_ARM_CONTROL = 5
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/base_env.py:45]: nit , typo : It should be `TRAVEL_TIME_JOINT_BASE_ARM_CONTROL`

Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/base_env.py:45]: @jimmytyyang @tusharsangam we already have a parameter in config.yaml :: `ARM_TRAJECTORY_TIME_IN_SECONDS`. Can we re-use that instead?

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 4]
```diff
@@ -335,11 +346,14 @@ def step(  # noqa
...
                 )  # / self.ctrl_hz
+                print("Slow down...")
             if base_action is not None and arm_action is not None:
+                print("input base_action velocity:", base_action)
+                print("input arm_action:", arm_action)
                 self.spot.set_base_vel_and_arm_pos(
                     *base_action,
                     arm_action,
-                    MAX_CMD_DURATION,
+                    travel_time=TRAVLE_TIME_JOINT_BASE_ARM_CONTROL,
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/envs/base_env.py:356]: Same as before

File: spot_rl_experiments/configs/config.yaml [hunk 5]
```diff
@@ -3,9 +3,13 @@ WEIGHTS:
   # GAZE: "weights/gaze_normal_32_seed300_1649708927_ckpt.49.pth"
   # PLACE: "weights/speed_sweep_seed2_speed0.174533_1648669786.ckpt.94.pth"
   # MIXER: "weights/sweep_17_SD_200_1651156250_ckpt.12.pth"
-# new
+
   # NAV: "weights/final_paper/nav_CUTOUT_WT_True_SD_300_ckpt.90.pth"
```
Comment (zephirefaith) [inline spot_rl_experiments/configs/config.yaml:7]: Can we clean up these commented out lines?

File: spot_rl_experiments/spot_rl/envs/gaze_env.py [hunk 6]
```diff
@@ -197,11 +232,34 @@ def step(self, base_action=None, arm_action=None, grasp=False, place=False):
...
+            place,
+            max_joint_movement_key="MAX_JOINT_MOVEMENT_MOBILE_GAZE"
+            if self._use_mobile_pick
+            else "MAX_JOINT_MOVEMENT",
+            max_lin_dist_mobile_gaze="MAX_LIN_DIST_MOBILE_GAZE"
+            if self._use_mobile_pick
+            else None,
+            max_ang_dist_mobile_gaze="MAX_ANG_DIST_MOBILE_GAZE"
+            if self._use_mobile_pick
+            else None,
```
Comment (tusharsangam) [inline spot_rl_experiments/spot_rl/envs/gaze_env.py:247]: Let's clean the function args @jimmytyyang

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 7]
```diff
@@ -335,11 +346,14 @@ def step(  # noqa
                 base_action = (
                     np.array(base_action) * self.slowdown_base
                 )  # / self.ctrl_hz
+                print("Slow down...")
```
Comment (zephirefaith) [inline spot_rl_experiments/spot_rl/envs/base_env.py:349]: Can we use loggers for this instead of print?

File: spot_rl_experiments/spot_rl/real_policy.py [hunk 8]
```diff
@@ -32,68 +47,133 @@ def __init__(
...
+                4,
+                512,
+                device=self.device,
+            )
 
         # We start an episode with 'done' being True (0 for 'not_done')
         self.not_done_masks = torch.zeros(1, 1, dtype=torch.bool, device=self.device)
         self.prev_actions = torch.zeros(1, self.num_actions, device=self.device)
 
+    def get_action(self, mu_maybe_std, std):
```
Comment (zephirefaith) [inline spot_rl_experiments/spot_rl/real_policy.py:134]: Confusing arg-name? `mu_maybe_std`?

File: spot_rl_experiments/spot_rl/real_policy.py [hunk 9]
```diff
@@ -296,7 +412,7 @@ def act(self, observations, expert=None):
 
 if __name__ == "__main__":
     gaze_policy = GazePolicy(
-        "weights/bbox_mask_5thresh_autograsp_shortrange_seed1_36.pth",
+        "weights/final_paper/gaze_normal_32_seed100_1649708902_ckpt.38.pth",
```
Comment (zephirefaith) [inline spot_rl_experiments/spot_rl/real_policy.py:415]: This should probably also come from a config or CLI

File: spot_rl_experiments/spot_rl/real_policy.py [hunk 10]
```diff
@@ -32,68 +47,133 @@ def __init__(
         action_space,
         device,
         policy_class=PointNavBaselinePolicy,
+        is_hab3_policy=False,
     ):
```
Comment (zephirefaith) [inline spot_rl_experiments/spot_rl/real_policy.py:51]: Can we get some class-level, module-level docstrings? It is hard to understand what is the purpose of this file.

File: spot_rl_experiments/utils/hab3_policy_conversion.py
Comment (zephirefaith) [inline spot_rl_experiments/utils/hab3_policy_conversion.py:1]: Same comment as previous file, some docstrings would significantly improve readability

File: tests/hardware_tests/test_mobile_gaze.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            4.979398852803741,
+            3.535594946585519,
+            np.deg2rad(np.rad2deg(-0.008869974097951427) + 15.0),
+        ),
+        "table_-10": (
+            4.979398852803741,
+            3.535594946585519,
+            np.deg2rad(np.rad2deg(-0.008869974097951427) - 15.0),
+        ),
+    }
```
Comment (zephirefaith) [inline tests/hardware_tests/test_mobile_gaze.py:35]: can these be read from a file or config instead?

**Review by zephirefaith — CHANGES_REQUESTED** [review]

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 12]
```diff
@@ -235,6 +236,8 @@ def step(  # noqa
         grasp=False,
         place=False,
         max_joint_movement_key="MAX_JOINT_MOVEMENT",
+        max_lin_dist_mobile_gaze=None,
+        max_ang_dist_mobile_gaze=None,
```
Comment (tusharsangam) [inline spot_rl_experiments/spot_rl/envs/base_env.py:240]: @jimmytyyang 
Maybe we can directly read this from YAML during initialization, as we dont need to change these variables on every iteration

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 13]
```diff
@@ -303,8 +308,14 @@ def step(  # noqa
...
-                lin_dist *= self.max_lin_dist
-                ang_dist *= self.max_ang_dist
+                if max_lin_dist_mobile_gaze is not None:
+                    lin_dist *= self.config[max_lin_dist_mobile_gaze]
+                else:
+                    lin_dist *= self.max_lin_dist
+                if max_ang_dist_mobile_gaze is not None:
+                    ang_dist *= self.config[max_ang_dist_mobile_gaze]
+                else:
+                    ang_dist *= self.max_ang_dist
```
Comment (tusharsangam) [inline spot_rl_experiments/spot_rl/envs/base_env.py:318]: This can be a class member variable defined inside init

File: spot_rl_experiments/spot_rl/real_policy.py [hunk 14]
```diff
@@ -32,68 +47,133 @@ def __init__(
...
+                self.config.RL.PPO.hidden_size,
+                device=self.device,
+            )
+        else:
+            self.test_recurrent_hidden_states = torch.zeros(
+                1,  # The number of environments. Just one for real world.
+                4,
+                512,
+                device=self.device,
+            )
```
Comment (tusharsangam) [inline spot_rl_experiments/spot_rl/real_policy.py:128]: @jimmytyyang 
let's not have magic numbers. You can move it to config.yaml

File: spot_rl_experiments/spot_rl/real_policy.py [hunk 15]
```diff
@@ -120,7 +200,43 @@ def __init__(self, checkpoint_path, device):
...
+                    low=0.0, high=1.0, shape=(240, 228, 1), dtype=np.float32
+                ),
+                "joint": spaces.Box(
+                    low=np.finfo(np.float32).min,
+                    high=np.finfo(np.float32).max,
+                    shape=(4,),
+                    dtype=np.float32,
+                ),
+            }
+        )
```
Comment (tusharsangam) [inline spot_rl_experiments/spot_rl/real_policy.py:231]: @jimmytyyang 
Can we test & check if we can use the same observation_space dict keys as that of static gaze ? Just so we don't want to remap then at run-time?

File: spot_rl_experiments/spot_rl/real_policy.py (hunk 9 above)
Comment (tusharsangam) [inline spot_rl_experiments/spot_rl/real_policy.py:415]: Let's revert this change as it is not tested

File: spot_rl_experiments/utils/hab3_policy_conversion.py [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+        return self.rsample(sample_shape)
+
+    def log_probs(self, actions) -> Tensor:
+        return super().log_prob(actions).sum(-1, keepdim=True)
+
+    def entropy(self) -> Tensor:
+        return super().entropy().sum(-1, keepdim=True)
+
+
+class RealPolicy:
```
Comment (tusharsangam) [inline spot_rl_experiments/utils/hab3_policy_conversion.py:43]:
```suggestion
class RealPolicyConverter:
```

File: spot_rl_experiments/utils/hab3_policy_conversion.py [hunk 17]
```diff
@@ -0,0 +1,341 @@
...
+        features = output_model[0]
+        rnn_hidden_states = output_model[1]
+
+        with torch.no_grad():
+            action = self.get_action(self.ts_action_dis(features), self.std)
+            action = action.mean
+        return action, rnn_hidden_states
+
+
+class MobileGazePolicy(RealPolicy):
```
Comment (tusharsangam) [inline spot_rl_experiments/utils/hab3_policy_conversion.py:195]:
```suggestion
class MobileGazePolicyConverter(RealPolicyConverter):
```

File: spot_rl_experiments/utils/hab3_policy_conversion.py [hunk 18]
```diff
@@ -0,0 +1,341 @@
...
+
+        features = output_model[0]
+        rnn_hidden_states = output_model[1]
+
+        with torch.no_grad():
+            action = self.get_action(self.ts_action_dis(features), self.std)
+            action = action.mean
+        return action, rnn_hidden_states
+
+
```
Comment (tusharsangam) [inline spot_rl_experiments/utils/hab3_policy_conversion.py:194]: @jimmytyyang 
Let's also add a documentation outlining the structure for all policy converter classes that will be defined in future.

File: tests/hardware_tests/test_mobile_gaze.py [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+        pick_from = pick_targets_keys[i]
+        i += 1
+        spotskillmanager = SpotSkillManager(use_mobile_pick=True)
+        spotskillmanager.nav(*pick_targets[pick_from])
+        x = input(f"Press Enter to continue to mobile gaze from {pick_from}")
+        spotskillmanager.pick("cereal_box")
+        spotskillmanager.spot.open_gripper()
+        contnue = map_user_input_to_boolean("Do you want to do it again ? Y/N ")
+
+    # Navigate to dock and shutdown
```
Comment (tusharsangam) [inline tests/hardware_tests/test_mobile_gaze.py:50]: Maybe we do not need this file. This can be added as a new function inside `test_gaze_env.py` called def `test_mobile_gaze()`

File: spot_rl_experiments/spot_rl/envs/base_env.py (hunk 7 above)
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/envs/base_env.py:349]: Replaced print with self.say method

File: spot_rl_experiments/configs/config.yaml (hunk 5 above)
Comment (tusharsangam) [reply spot_rl_experiments/configs/config.yaml:7]: done @zephirefaith

File: spot_rl_experiments/spot_rl/real_policy.py (hunk 8 above)
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/real_policy.py:134]: @jimmytyyang  might be able to comment on this, if its okay then I shall rename mu_maybe_std with mu

File: spot_rl_experiments/spot_rl/real_policy.py (hunk 10 above)
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/real_policy.py:51]: Posted this on slack channel

File: spot_rl_experiments/utils/hab3_policy_conversion.py [hunk 20]
```diff
@@ -0,0 +1,341 @@
...
+    # Get the observation space
+    if args.use_stereo_pair_camera:
+        observations = {
+            "arm_depth_bbox_sensor": np.zeros([240, 228, 1], dtype=np.float32),
+            "articulated_agent_arm_depth": np.zeros([240, 228, 1], dtype=np.float32),
+            "spot_head_stereo_depth_sensor": np.zeros([240, 228, 1], dtype=np.float32),
+            "joint": np.zeros(4, dtype=np.float32),
+        }
+    else:
+        observations = {
```
[the remark this answers is not in the document]
Comment (tusharsangam) [reply spot_rl_experiments/utils/hab3_policy_conversion.py:316]: Changes done

File: spot_rl_experiments/spot_rl/real_policy.py (hunk 9 above)
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/real_policy.py:415]: changed to load from config @zephirefaith , will reflect in next commit

File: tests/hardware_tests/test_mobile_gaze.py (hunk 11 above)
Comment (tusharsangam) [reply tests/hardware_tests/test_mobile_gaze.py:35]: @zephirefaith  addressed, will be reflected in next commit, moved these nav targets in waypoints.yaml

File: spot_rl_experiments/spot_rl/envs/base_env.py (hunk 13 above)
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/envs/base_env.py:318]: @jimmytyyang

File: spot_rl_experiments/spot_rl/real_policy.py (hunk 14 above)
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/real_policy.py:128]: @jimmytyyang addressed it, will be loaded from config & will be seen in next commit

File: spot_rl_experiments/utils/hab3_policy_conversion.py (hunk 16 above)
Comment (tusharsangam) [reply spot_rl_experiments/utils/hab3_policy_conversion.py:43]: addressed

File: spot_rl_experiments/utils/hab3_policy_conversion.py (hunk 18 above)
Comment (tusharsangam) [reply spot_rl_experiments/utils/hab3_policy_conversion.py:194]: @jimmytyyang  addressed , will reflect in next commit

File: tests/hardware_tests/test_mobile_gaze.py (hunk 19 above)
Comment (tusharsangam) [reply tests/hardware_tests/test_mobile_gaze.py:50]: @KavitShah1998  this is not a good idea as test_gaze_env.py is designed to test GazeController instead of skillmanager

**Comment by jimmytyyang** [comment]
@akshararai  this is the PR for mobile gaze

File: spot_rl_experiments/spot_rl/envs/gaze_env.py (hunk 6 above)
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/gaze_env.py:237]: okay, fixed this! thank you!

File: spot_rl_experiments/spot_rl/envs/base_env.py (hunk 12 above)
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/base_env.py:239]: okay, fixed this! thank you!

File: spot_rl_experiments/spot_rl/envs/base_env.py (hunk 13 above)
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/base_env.py:319]: okay, fixed this! thank you!

File: spot_rl_experiments/spot_rl/real_policy.py (hunk 8 above)
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/real_policy.py]: Hi all, this is to follow the convention from hab3

File: spot_rl_experiments/spot_rl/real_policy.py (hunk 15 above)
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/real_policy.py]: Hi @KavitShah1998 @tusharsangam! After investigation, the name for the observation keys needs to be the same as the one when we do conversion from Hab3 to torchscript. Since when do conversion, we need to use the exact name when we train the policy to do conversion, it is not possible to change the name of observation keys. And I think we should use the consistent name for observations nevertheless since we make sure there is no confusion

File: spot_rl_experiments/spot_rl/real_policy.py [hunk 21]
```diff
@@ -24,76 +29,160 @@ def to_tensor(v):
...
+                    self.test_recurrent_hidden_states,
+                    self.prev_actions,
+                    self.not_done_masks,
+                    deterministic=True,
+                    actions_only=True,
+                )
+            else:
+                # Using torch script to save the model
+                with torch.no_grad():
+                    output_model = self.policy["net"](
```
Comment (akshararai) [inline spot_rl_experiments/spot_rl/real_policy.py:170]: why is this model loaded every policy call? Should this be loaded once and then used here?

File: spot_rl_experiments/utils/pytorch_to_torchscript.py
Comment (akshararai) [inline spot_rl_experiments/utils/pytorch_to_torchscript.py:1]: if this file is based on another source, please add a citation to it.

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 22]
```diff
@@ -246,29 +244,33 @@ def step(  # noqa
...
         :return: observations, reward (None), done, info
         """
         assert self.reset_ran, ".reset() must be called first!"
         target_yaw = None
         if disable_oa is None:
             disable_oa = self.config.DISABLE_OBSTACLE_AVOIDANCE
         grasp = grasp or self.config.GRASP_EVERY_STEP
-        print(f"raw_base_ac: {arr2str(base_action)}\traw_arm_ac: {arr2str(arm_action)}")
+        self.say(
+            f"raw_base_ac: {arr2str(base_action)}\traw_arm_ac: {arr2str(arm_action)}"
```
[the remark this answers is not in the document]
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/envs/base_env.py:257]: I will add that @akshararai

File: spot_rl_experiments/utils/pytorch_to_torchscript.py
Comment (tusharsangam) [reply spot_rl_experiments/utils/pytorch_to_torchscript.py:1]: @akshararai we took inspiration from torchscript tutorial from pytorch that is just the torch.jit.trace part other than that its original work from @jimmytyyang

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 23]
```diff
@@ -246,29 +244,33 @@ def step(  # noqa
         :param place: whether to call the open_gripper() method
         :param max_joint_movement_key: max allowable displacement of arm joints
             (different for gaze and place)
+        :param max_lin_dist_key: maximum linear distance allowed if sepecify
```
[the remark this answers is not in the document]
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/envs/base_env.py:247]: changes incorporated

File: spot_rl_experiments/spot_rl/real_policy.py [hunk 24]
```diff
@@ -24,76 +29,160 @@ def to_tensor(v):
...
+                _, actions, _, self.test_recurrent_hidden_states = self.policy.act(
+                    batch,
+                    self.test_recurrent_hidden_states,
+                    self.prev_actions,
+                    self.not_done_masks,
+                    deterministic=True,
+                    actions_only=True,
+                )
+            else:
+                # Using torch script to save the model
```
[the remark this answers is not in the document]
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/real_policy.py:168]: incorporated

File: spot_rl_experiments/spot_rl/real_policy.py (hunk 21 above)
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/real_policy.py:170]: @akshararai the variable name might be confusing, its output_of_the_model & not model itself. As you said, this model is loaded only once.

File: spot_rl_experiments/spot_rl/real_policy.py [hunk 25]
```diff
@@ -24,76 +29,160 @@ def to_tensor(v):
...
         with torch.no_grad():
-            _, actions, _, self.test_recurrent_hidden_states = self.policy.act(
-                batch,
-                self.test_recurrent_hidden_states,
-                self.prev_actions,
-                self.not_done_masks,
-                deterministic=True,
-                actions_only=True,
-            )
+            if not self.is_hab3_policy:
```
[the remark this answers is not in the document]
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/real_policy.py:158]: Yes that's good idea, however do you want to incorporate that in this PR or there should be another PR after all the PRs are merged ?

[the remark this answers is not in the document]
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/real_policy.py:158]: Let's have a separate PR for this. I have documented this! Thanks for the suggestion!

File: spot_rl_experiments/configs/config.yaml [hunk 26]
```diff
@@ -27,25 +35,36 @@ SUCCESS_ANGLE_DIST: 5
...
+PLACE_ACTION_SPACE_LENGTH: 4
 
 # Base action params
 MAX_LIN_DIST: 0.25  # meters
 MAX_ANG_DIST: 15.0  # degrees
 
 # Arm action params
 MAX_JOINT_MOVEMENT: 0.0872665  # Gaze arm speed (5 deg)
-MAX_JOINT_MOVEMENT_2: 0.174533  # Place arm speed (6 deg)
+MAX_JOINT_MOVEMENT_2: 0.174533  # Place arm speed (10 deg)
```
[the remark this answers is not in the document]
Comment (jimmytyyang) [reply spot_rl_experiments/configs/config.yaml:67]: Thanks, I will document this, and have another PR to fix this!

File: spot_rl_experiments/spot_rl/real_policy.py [hunk 27]
```diff
@@ -24,76 +29,160 @@ def to_tensor(v):
...
 
         # We start an episode with 'done' being True (0 for 'not_done')
         self.not_done_masks = torch.zeros(1, 1, dtype=torch.bool, device=self.device)
         self.prev_actions = torch.zeros(1, self.num_actions, device=self.device)
 
+    def get_action_distribution(self, mu_maybe_std, std):
+        """The final transformation of the action given inputs"""
+        # We use mu_maybe_std to follow the convension of habitat-baselines
+        mu = torch.tanh(mu_maybe_std.float())
+        std = torch.exp(torch.clamp(std, -5, 2))
```
[the remark this answers is not in the document]
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/real_policy.py:150]: @jimmytyyang

File: spot_rl_experiments/spot_rl/envs/gaze_env.py [hunk 28]
```diff
@@ -197,11 +222,38 @@ def step(self, base_action=None, arm_action=None, grasp=False, place=False):
         grasp = self.should_grasp()
 
         observations, reward, done, info = super().step(
-            base_action, arm_action, grasp, place
+            base_action,
+            arm_action,
+            grasp,
+            place,
+            max_joint_movement_key="MAX_JOINT_MOVEMENT_MOBILE_GAZE"
```
[the remark this answers is not in the document]
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/gaze_env.py:229]: @jimmytyyang

File: spot_rl_experiments/spot_rl/envs/base_env.py [hunk 29]
```diff
@@ -303,8 +306,11 @@ def step(  # noqa
             if np.count_nonzero(base_action) > 0:
                 # Command velocities using the input action
                 lin_dist, ang_dist = base_action
-                lin_dist *= self.max_lin_dist
-                ang_dist *= self.max_ang_dist
+
+                # Scale the linear and angular velocities
+                lin_dist *= self.config[max_lin_dist_key]
```
[the remark this answers is not in the document]
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/base_env.py:312]: @jimmytyyang

File: spot_rl_experiments/spot_rl/real_policy.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/real_policy.py:150]: Thank you, I moved these to config.

File: spot_rl_experiments/spot_rl/envs/gaze_env.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/gaze_env.py:229]: Done, thank you!

File: spot_rl_experiments/spot_rl/envs/base_env.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/base_env.py:312]: Thanks for this! We are not using hydra to look up the config; instead, the config represents a dictionary, which should be fast. I agree that we should cache the values for `max_joint_movement_key`, `max_lin_dist_key`, and `max_ang_dist_key` when we initialize the `env`. I have addressed this by caching the value during initialization.

File: spot_rl_experiments/spot_rl/real_policy.py [hunk 30]
```diff
@@ -24,76 +29,160 @@ def to_tensor(v):
...
+            self.test_recurrent_hidden_states = torch.zeros(
+                1,  # The number of environments. Just one for real world.
+                self.policy.net.num_recurrent_layers,
+                self.config.RL.PPO.hidden_size,
+                device=self.device,
+            )
+        else:
+            self.test_recurrent_hidden_states = torch.zeros(
+                1,  # The number of environments. Just one for real world.
+                self.config.get("MOBILE_GAZE_NUM_RECURRENT_LAYERS", 4),
```
[the remark this answers is not in the document]
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/real_policy.py]: Can't store attributes in torchscipt model as we are using torch.jit.trace method instead of torch.jit. script method for converting pytorch model into torchscript. Currently torch.jit.script method is throwing error & thus without breaking anything, the simplest solution I came up with is to store these attributes in standard deviation's pth file.
I created a dictionary with std tensor & other required attributes before saving to .pth file (standard deviation tensor is loaded separately during inference).    @akshararai

**Review by jimmytyyang — APPROVED** [review]
LGTM!

**Comment by jimmytyyang** [comment]
@zephirefaith Hi Priyam! Tushar and I have addressed all the awesome comments from all the reviewers. Could you please take a look at it if you have time? I think this PR is clean right now, and we would like to merge this PR by today. Thank you so much for your contribution!

**Comment by tusharsangam** [comment]
@zephirefaith I have moved all the waypoints in the waypoints.yaml file.
